File: 0539. Minimum Time Difference/Solution0539.cpp
#include<iostream>
#include<vector>
#include<algorithm>
using namespace std;

class Solution0539 { // 18 - 24
public:
    int findMinDifference(vector<string>& timePoints) {
        int n = timePoints.size();
        vector<int> times(n,0);
        for (int i = 0; i < n; i++) {
            times[i] = convTime(timePoints[i]);
        }
        sort(times.begin(), times.end());
        int res = 24*60+1;
        for (int i = 1; i < n; i++) {
            res = min(times[i]-times[i-1], res);
        }
        res = min(res, times[n-1]-times[0]);        
        res = min(res, times[0]+24*60-times[n-1]);
        return res;
    }
    
    int convTime(string timePoint) {
        int hr = stoi(timePoint.substr(0,2));        
        int min = stoi(timePoint.substr(3));
        return hr * 60 + min;
    }
};

File: 0539. Minimum Time Difference/Solution0539.py
class Solution0539:
    def findMinDifference(self, timePoints: List[str]) -> int:
        t = sorted(int(i[:2])*60+int(i[-2:]) for i in timePoints)
        t.append(t[0]+24*60)
        return min(b-a for a, b in zip(t, t[1:]))

    def findMinDifference0(self, timePoints: List[str]) -> int:
        def convTime(timePoint: str) -> int:
            hr, mn = int(timePoint[:2]), int(timePoint[3:])
            return hr*60+mn 
            
        n,res = len(timePoints), 24*60    
        times = []
        for timePoint in timePoints:
            times.append(convTime(timePoint))
        times.sort()
        for i in range(1,n):
            res = min(res, times[i]-times[i-1])
        res = min(res, times[0]+24*60-times[n-1])
        return res